Clases/man/man3/FunctionsIO.hh.3
.TH "FunctionsIO.hh" 3 "Miércoles, 7 de Diciembre de 2016" "Version v1.1" "Calculadora de expresiones aritméticas" \" -*- nroff -*-
.ad l
.nh
.SH NAME
FunctionsIO.hh \- Especificación de las operaciones de lectura/escritura de expresiones\&.  

.SH SYNOPSIS
.br
.PP
\fC#include 'Environment\&.hh'\fP
.br
\fC#include 'Expression\&.hh'\fP
.br

.SS "Funciones"

.in +1c
.ti -1c
.RI "void \fBgetString\fP (string &str)"
.br
.RI "\fIOperación de separación de strings\&. \fP"
.ti -1c
.RI "bool \fBreadExpression\fP (\fBEnvironment\fP &env, \fBExpression\fP &exp)"
.br
.RI "\fIOperación de lectura de expresiones\&. \fP"
.ti -1c
.RI "void \fBwriteExpression\fP (const \fBEnvironment\fP &env, const \fBExpression\fP &exp)"
.br
.RI "\fIOperación de escritura de expresiones\&. \fP"
.in -1c
.SH "Descripción detallada"
.PP 
Especificación de las operaciones de lectura/escritura de expresiones\&. 


.SH "Documentación de las funciones"
.PP 
.SS "void getString (string & str)"

.PP
Operación de separación de strings\&. 
.PP
\fBPrecondición:\fP
.RS 4
\fICierto\fP 
.RE
.PP
\fBPostcondición:\fP
.RS 4
'str' es un string contenido entre espacios o paréntesis en la entrada que se ha leído 
.RE
.PP

.SS "bool readExpression (\fBEnvironment\fP & env, \fBExpression\fP & exp)"

.PP
Operación de lectura de expresiones\&. 
.PP
\fBPrecondición:\fP
.RS 4
'env' es un entorno con un espacio de operaciones primitivas inicializado con las operaciones primitivas predefinidas; 'exp' es una expresión vacía 
.RE
.PP
\fBPostcondición:\fP
.RS 4
Se han leído los datos del canal estándar de entrada y se han asignado a 'env' si eran una definición de una variable o operación o a 'exp' si representaban una expresión evaluable 
.RE
.PP

.SS "void writeExpression (const \fBEnvironment\fP & env, const \fBExpression\fP & exp)"

.PP
Operación de escritura de expresiones\&. 
.PP
\fBPrecondición:\fP
.RS 4
'env' es un entorno con un espacio de operaciones primitivas inicializado con las operaciones primitivas predefinidas; 'exp' es una expresión no vacía 
.RE
.PP
\fBPostcondición:\fP
.RS 4
Escribe el valor de la expresión representada por 'exp' por el canal estándar de salida 
.RE
.PP

.SH "Autor"
.PP 
Generado automáticamente por Doxygen para Calculadora de expresiones aritméticas del código fuente\&.


Code/FunctionsIO.cc
/** @file FunctionIO.cc
    @brief Implementación del módulo de operaciones de entrada/salida
*/

#ifndef NO_DIAGRAM
#include <iostream>
#include <string>
#endif

#include "FunctionsIO.hh"
#include "Input.hh"
#include "Environment.hh"
#include "Expression.hh"

using namespace std;

typedef list<Expression*>::const_iterator const_iter;


bool is_num(string str)
{
	int i = 0;
	if(not str.empty() and (str[0] == '-' or (str[0] >= '0' and str[i] <= '9'))) {
			++i;
		/* INV
			0 <= i < str.size()
		*/
		while(i < str.size() and str[i] >= '0' and str[i] <= '9') {
			++i;
		}
	}
	return not str.empty() and i == str.size() and str[i-1] != '-';
}

void createLocalVarSpace(Environment& env, Expression& exp, string parameters)
{
	Input in;
	in << parameters;
	/* INV
		c_it es un iterador a un elemento de la lista de expresiones de 'exp' comprendido entre exp.begin() y exp.end()
	*/
	for(const_iter c_it = exp.begin(); c_it != exp.end(); ++c_it) {
		env.set_var(++in, *(*c_it));
	}
}

void evaluate(Environment& env, Expression& exp)
{
	if(exp.is_op()){
		if(env.is_primitive(exp.get_op())) {
			env.get_prim(exp.get_op())(exp);
		}
		else if(env.exists_op(exp.get_op())) {
			string parameters = env.get_op(exp.get_op()).first;
			string express = env.get_op(exp.get_op()).second;
			Environment lEnv(env);
			lEnv.erase_varspace();
			if(exp.size() == env.num_pars_op(exp.get_op())){
				createLocalVarSpace(lEnv, exp, parameters);
				Input in;
				in << express;
				++in;
				exp.clear();
				readExpression(lEnv, exp, in);
			}
			else {
				exp.set_undefined();
			}
		}
	}
	else if(exp.is_list()) {
		const_iter c_it = exp.begin();
		/* INV
			c_it es un iterador a un elemento de la lista de expresiones de 'exp' comprendido entre exp.begin() y exp.end()
		*/
		while(c_it != exp.end() and (*c_it) -> is_value()) {
			++c_it;
		}
		if(c_it != exp.end()) {
			exp.set_undefined();
		}
	}
}

void readList(Environment& env, Expression& exp, Input& in)
{
	if(in.get_str() != ")") {
		Expression* pExp;
		/* INV
			'pExp' es un puntero a una expresión a ser insertada al final de la lista de expresiones de 'exp'
		*/
		do {
			pExp = new Expression;
			readExpression(env, *pExp, in);
			exp.insert(exp.end(), pExp);
		} while(++in != ")" and not pExp -> undefined());
		if(pExp -> undefined()) {
			exp.set_undefined();
		}
		else {
			exp.set_list();
		}
	}
	else {
		exp.set_empty_list();
	}
}

void readIf(Environment& env, Expression& exp, Input& in)
{
	string str;
	Expression* pExp = new Expression;
	readExpression(env, *pExp, in);
	exp.insert(exp.end(), pExp);
	if(pExp -> is_bool()){
		int b = pExp -> get_value();
		if(b == 0) {
			in.nextExpression();
		}
		if(in.get_p() > 0) {
			++in;
			pExp = new Expression;
			readExpression(env, *pExp, in);
			exp.insert(exp.end(), pExp);
			if(pExp -> undefined() or  (2 - b + in.countExpressions() != 2)) {
				exp.set_undefined();
			}
		}
		else {
			exp.set_undefined();
		}
	}
	else {
		exp.set_undefined();
	}
}

void readOp(Environment& env, Expression& exp, Input& in)
{
	exp.set_op(in.get_str());
	++in;
	if(exp.get_op() == "if") {
		readIf(env, exp, in);
	}
	else {
		Expression aux;
		readList(env, aux, in);
		if(aux.undefined()) {
			exp.set_undefined();
		}
		else {
			exp.swap_list(aux);
		}
	}
}

void defineVar(Environment& env, string key, Input& in)
{
	Expression exp;
	readExpression(env, exp, in);
	if(not exp.undefined()) {
		env.set_var(key, exp);
		env.write_var(key);
	}
	else {
		cout << key << " " << "indefinit" << endl;
	}
}

void defineOp(Environment& env, string key, string parameters, Input& in)
{
	--in;
	env.set_op(key, parameters, in.nextExpression());
	env.write_op(key);
}

void define(Environment& env, Input& in)
{
	string key = ++in;
	string aux = in.nextExpression();
	++in;
	if(in.get_p() == 0) {
		in << aux;
		++in;
		defineVar(env, key, in);
	}
	else{
		if(aux.size() < 4) {
			aux.clear();
		}
		else {
			aux.erase(0, 2);
			aux.erase(aux.size()-2, 2);
		}
		defineOp(env, key, aux, in);
	}
}

bool readExpression(Environment& env, Expression& exp, Input& in)
{
	string str = in.get_str();
	if(str == "(") {
		in >> str;
		if(env.is_op(str)) {
			readOp(env, exp, in);
		}
		else if(str == "define") {
			define(env, in);
		}
		else {
			readList(env, exp, in);
		}
	}
	else if(is_num(str)){
		exp.set_value(stoi(str));
	}
	else if(env.exists_var(str)) {
		exp = env.get_var(str);
	}
	else {
		exp.set_undefined();
	}
	evaluate(env, exp);
	return str != "define";
}

bool read(Environment& env, Expression& exp)
{
	Input in;
	string str = ++in;
	if(str != "****" and readExpression(env, exp, in)){
		exp.write();
	}
	in.endOfExpression();
	return str != "****";
}
